src/core/utils/components/VirtualScroll.tsx

```diff
@@ -100,145 +131,232 @@ export type VirtualScrollProps = {
-      ? parentRef.current?.ownerDocument.removeEventListener(
-          'scroll',
-          onScrollRef.current,
-        )
-      : scrollContainer.current.removeEventListener(
-          'scroll',
-          onScrollRef.current,
-        );
-  }, []);
-
-  // Add event listener to the scrollable container.
-  React.useLayoutEffect(() => {
-    removeScrollListener();
-    onScrollRef.current = updateVirtualScroll;
-    if (
+    const removeScrollListener = React.useCallback(() => {
+      if (!onScrollRef.current) {
+        return;
+      }
       !scrollContainer.current ||
       scrollContainer.current === parentRef.current?.ownerDocument.body
-    ) {
-      parentRef.current?.ownerDocument.addEventListener(
-        'scroll',
-        updateVirtualScroll,
-      );
-    } else {
-      scrollContainer.current.addEventListener('scroll', updateVirtualScroll);
-    }
-    return removeScrollListener;
-  }, [updateVirtualScroll, removeScrollListener]);
-
-  React.useLayoutEffect(() => {
-    updateVirtualScroll();
-  }, [scrollContainerHeight, itemsLength, updateVirtualScroll]);
-
-  return (
-    <div
-      style={{
-        overflow: 'hidden',
-        minHeight: itemsLength * childHeight.current,
-        width: '100%',
-        ...style,
-      }}
-      ref={ref}
-      {...rest}
-    >
+        ? parentRef.current?.ownerDocument.removeEventListener(
+            'scroll',
+            onScrollRef.current,
+          )
+        : scrollContainer.current.removeEventListener(
+            'scroll',
+            onScrollRef.current,
+          );
+    }, []);
+
+    // Add event listener to the scrollable container.
+    React.useLayoutEffect(() => {
+      removeScrollListener();
+      onScrollRef.current = onScroll;
+      if (
+        !scrollContainer.current ||
+        scrollContainer.current === parentRef.current?.ownerDocument.body
+      ) {
+        parentRef.current?.ownerDocument.addEventListener('scroll', onScroll);
+      } else {
+        scrollContainer.current.addEventListener('scroll', onScroll);
+      }
+      return removeScrollListener;
+    }, [onScroll, removeScrollListener]);
+
+    React.useLayoutEffect(() => {
+      if (!isMounted) {
+        return;
+      }
+
+      const scrollableContainer = getScrollableContainer();
+
+      // if `scrollToIndex` is not visible, scroll to it
+      if (
+        scrollableContainer &&
+        scrollToIndex != null &&
+        (scrollToIndex > visibleIndex.current.end ||
+          scrollToIndex < visibleIndex.current.start)
+      ) {
+        const indexDiff =
+          scrollToIndex > visibleIndex.current.end
+            ? scrollToIndex - visibleIndex.current.end
+            : scrollToIndex - visibleIndex.current.start;
+
+        // scrollTop value to display full item height
+        const adjustedScrollTop =
+          scrollableContainer.scrollTop -
+          (scrollableContainer.scrollTop % childHeight.current.middle);
```

Comment:
What does this calculation do? Not sure I understand.

Reply:
If at that time, scroll is somehow in the middle (displaying not full height element), I want to adjust the `scrollTop` value so I scroll the amount of pixels, that first element is displayed fully (not in the middle or just a part of it)

![image](https://user-images.githubusercontent.com/81580355/157024154-165975fe-c003-4cc4-9546-548ce16aa6ec.png)

Reply:
Have a little different now.. Read, if its readable :D